File: wrappers/fedimint-py/models/lightning.py
from pydantic import BaseModel
from typing import Any, List, Optional, Union
from enum import Enum


class LightningCreateInvoiceRequest(BaseModel):
    amountMsat: int
    description: str
    expiryTime: Optional[int]


class LightningCreateInvoiceResponse(BaseModel):
    operationId: str
    invoice: str


class LightningInvoiceForPubkeyTweakRequest(BaseModel):
    amountMsat: int
    description: str
    externalPubkey: str
    tweak: int
    expiryTime: Optional[int]


class LightningInvoiceForPubkeyTweakResponse(BaseModel):
    operationId: str
    invoice: str


class LightningClaimPubkeReceivesRequest(BaseModel):
    privateKey: str
    tweaks: List[int]


class LightningAwaitInvoiceRequest(BaseModel):
    operationId: str


class LightningPayRequest(BaseModel):
    paymentInfo: str
    amountMsat: Optional[int]
    lightningUrlComment: Optional[str]


class LightningPayResponse(BaseModel):
    operationId: str
    paymentType: str
    contractId: str
    fee: int


class LightningAwaitPayRequest(BaseModel):
    operationId: str


class LnReceiveState(Enum):
    Created = "Created"
    WaitingForPayment = "WaitingForPayment"
    Canceled = "Canceled"
    Funded = "Funded"
    AwaitingFunds = "AwaitingFunds"
    Claimed = "Claimed"


class WaitingForPayment(BaseModel):
    invoice: str
    timeout: int


class Canceled(BaseModel):
    reason: str


class LightningPaymentResponse(BaseModel):
    state: LnReceiveState
    details: Optional[Union[WaitingForPayment, Canceled]] = None


class GatewayFees(BaseModel):
    baseMsat: int
    proportionalMillionths: int


class GatewayInfo(BaseModel):
    api: str
    fees: GatewayFees
    gatewayId: str
    gatewayRedeemKey: str
    lightningAlias: str
    mintChannelId: int
    nodePubKey: str
    routeHints: List[Any]
    supportsPrivatePayments: bool


class GatewayTTL(BaseModel):
    nanos: int
    secs: int


class Gateway(BaseModel):
    federationId: str
    info: GatewayInfo
    ttl: GatewayTTL
    vetted: bool

File: wrappers/fedimint-kotlin/src/main/kotlin/com/fedimintClientd/models/Lightning.kt
package com.fedimintClientd.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class LightningCreateInvoiceResponse(
    val operationId: String,
    val invoice: String
)

@Serializable
data class LightningInvoiceForPubkeyTweakResponse(
    val operationId: String,
    val invoice: String
)

@Serializable
data class LightningPayResponse(
    val operationId: String,
    val paymentType: PaymentType,
    val contractId: String,
    val fee: Int
)

@Serializable
data class PaymentType(
    val internal: String,
)

@Serializable
sealed class LnPaymentResult {
    data class WaitingForPayment(val paymentRequest: String) : LnPaymentResult()
    object Canceled : LnPaymentResult()
}

@Serializable
data class LightningPaymentResponse(
    val state: LnReceiveState,
    val details: LnPaymentResult? = null
)

@Serializable
enum class LnReceiveState {
    Created,
    WaitingForPayment,
    Canceled,
    Funded,
    AwaitingFunds,
    Claimed
}

@Serializable
data class GatewayFees(
    @SerialName("base_msat")
    val baseMsat: Int,
    @SerialName("proportional_millionths")
    val proportionalMillionths: Int
)

@Serializable
data class GatewayInfo(
    val api: String,
    val fees: GatewayFees,
    @SerialName("gateway_id")
    val gatewayId: String,
    @SerialName("gateway_redeem_key")
    val gatewayRedeemKey: String,
    @SerialName("lightning_alias")
    val lightningAlias: String,
    @SerialName("mint_channel_id")
    val mintChannelId: Int,
    @SerialName("node_pub_key")
    val nodePubKey: String,
    @SerialName("route_hints")
    val routeHints: List<String>,
    @SerialName("supports_private_payments")
    val supportsPrivatePayments: Boolean
)

@Serializable
data class GatewayTTL(
    val nanos: Long,
    val secs: Int
)

@Serializable
data class Gateway(
    @SerialName("federation_id")
    val federationId: String,
    val info: GatewayInfo,
    val ttl: GatewayTTL,
    val vetted: Boolean
)
